### PR TITLE
Phone number changing

### PR DESCRIPTION
**Pull request by vytis**
## What's in this PR?

Seems to be quite useful to be able to edit your phone number and Android had that for ages!

- [x] First merge in #671 

EDIT: Needs a release with https://github.com/wireapp/wire-ios-sync-engine/commit/9c0139de43a5266fa72577bf7a6dff92847d9ca2

### CODE REVIEW

File: Wire-iOS/Sources/UserInterface/Settings/ChangingDetails/ConfirmPhoneViewController.swift [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        tableView.deselectRow(at: indexPath, animated: false)
+    }
+}
+
+extension ConfirmPhoneViewController: ZMUserObserver {
+    
+    func userDidChange(_ note: ZMCDataModel.UserChangeInfo!) {
+        if note.user.isSelfUser {
+            showLoadingView = false
+            // we need to check if the notification really happened because
```
Comment (daehn) [inline Wire-iOS/Sources/UserInterface/Settings/ChangingDetails/ConfirmPhoneViewController.swift]: If the change is unrelated to the phone number we might hide the loading view to early or not?

File: Wire-iOS/Sources/UserInterface/Settings/ChangingDetails/ChangePhoneViewController.swift [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+        
+    }
+    
+    func validate() -> ValidationResult {
+        var validatedNumber = fullNumber as NSString?
+        let pointer = AutoreleasingUnsafeMutablePointer<NSString?>(&validatedNumber)
+        do {
+            try ZMUser.validatePhoneNumber(pointer)
+        } catch let error as NSError {
+            guard let errorCode = ZMManagedObjectValidationErrorCode(rawValue: UInt(error.code)) else { return .invalid }
```
Comment (mikeger) [inline Wire-iOS/Sources/UserInterface/Settings/ChangingDetails/ChangePhoneViewController.swift]: Would be cool to have a fail-able initializer on `ValidationResult` that takes `NSError`

File: Wire-iOS/Sources/UserInterface/Settings/ChangingDetails/ChangePhoneViewController.swift [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+fileprivate struct ChangePhoneNumberState {
+    let currentNumber: PhoneNumber?
+    var newNumber: PhoneNumber?
+    
+    var visibleNumber: PhoneNumber? {
+        return newNumber ?? currentNumber
+    }
+    
+    var saveButtonEnabled: Bool {
```
Comment (mikeger) [inline Wire-iOS/Sources/UserInterface/Settings/ChangingDetails/ChangePhoneViewController.swift]: I would rename to `canBeSaved` or `isValid`, since `ChangePhoneNumberState` is not really a view state. Just a suggestion.

File: Wire-iOS/Sources/UserInterface/Settings/ChangingDetails/ChangePhoneViewController.swift (hunk 2 above)
Comment (vytis) [inline Wire-iOS/Sources/UserInterface/Settings/ChangingDetails/ChangePhoneViewController.swift]: Good idea, makes the code much more straightforward

File: Wire-iOS/Sources/UserInterface/Settings/ChangingDetails/ChangePhoneViewController.swift (hunk 3 above)
Comment (vytis) [inline Wire-iOS/Sources/UserInterface/Settings/ChangingDetails/ChangePhoneViewController.swift]: True, renamed.

**Comment by mikeger** [comment]
💣  Conflict (rather small one). BTW when this one can be merged?

**Comment by jenskasemets** [comment]
@mikeger I think this can be merged now

**Comment by vytis** [comment]
@jenskasemets it actually needs release from zmessaging to work - which is not yet done as some tests are failing

**Comment by mikeger** [comment]
@vytis 1. Conflict 2. Can be merged now?
